core/drand_proxy.go

```diff
@@ -14,13 +17,113 @@ type drandProxy struct {
 	r drand.PublicServer
 }
 
+// streamProxy directly relays mesages of the PublicRandResponse stream.
+type streamProxy struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	incoming chan *drand.PublicRandResponse
+	outgoing chan *drand.PublicRandResponse
+}
+
+func newStreamProxy(ctx context.Context) *streamProxy {
+	ctx, cancel := context.WithCancel(ctx)
+	s := streamProxy{
+		ctx:      ctx,
+		cancel:   cancel,
+		incoming: make(chan *drand.PublicRandResponse, 0),
+		outgoing: make(chan *drand.PublicRandResponse, 1),
+	}
+	go s.loop()
+	return &s
+}
+
+func (s *streamProxy) Recv() (*drand.PublicRandResponse, error) {
+	next, ok := <-s.outgoing
```

Comment:
Select with `s.ctx`?

Reply:
the other side of the channel `loop()` closes it on context close. i'll leave a comment.